siva/checkpoint.go

```diff
@@ -0,0 +1,150 @@
+package siva
+
+import (
+	"os"
+	"strconv"
+
+	borges "github.com/src-d/go-borges"
+	billy "gopkg.in/src-d/go-billy.v4"
+	"gopkg.in/src-d/go-billy.v4/util"
+	errors "gopkg.in/src-d/go-errors.v1"
+)
+
+var (
+	// ErrCannotUseCheckpointFile is returned on checkpoint problems.
+	ErrCannotUseCheckpointFile = errors.NewKind("cannot use checkpoint file: %s")
+	// ErrCannotUseSivaFile is returned on siva problems.
+	ErrCannotUseSivaFile = errors.NewKind("cannot use siva file: %s")
+)
+
+const checkpointExtension = ".checkpoint"
+
+// Checkpoint tracks the status of a siva file and creates checkpoints to be
+// able to return back to a known state of that siva file.
+type Checkpoint struct {
+	offset  int64
+	baseFs  billy.Filesystem
+	path    string
+	persist string
+}
+
+// NewCheckpoint builds a new Checkpoint.
+func NewCheckpoint(fs billy.Filesystem, path string) (*Checkpoint, error) {
+	persist := path + checkpointExtension
+
+	if _, err := fs.Stat(path); err != nil && os.IsNotExist(err) {
+		cleanup(fs, persist)
+		return nil, ErrCannotUseSivaFile.Wrap(
+			borges.ErrLocationNotExists.New(path))
+	}
+
+	c := &Checkpoint{
+		baseFs:  fs,
+		path:    path,
+		persist: persist,
+	}
+
+	offset, err := readInt64(fs, persist)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, ErrCannotUseCheckpointFile.Wrap(err)
+		}
+
+		offset = 0
+	}
+
+	c.offset = offset
+
+	if err := c.Apply(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Apply applies if necessary the operations on the siva file to
+// leave it in the last correct state the checkpoint keeps.
+func (c *Checkpoint) Apply() error {
+	if c.offset > 0 {
```

Comment:
I would set offset to a negative number as there's a case not done here. When we start a transaction with a siva file that does not exist, the checkpoint will contain `0`. This means that if we find a checkpoint with offset `0` the file must be deleted.